apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -22,6 +22,520 @@ import (
+
+	// Envoy contains parameters for Envoy as well
+	// as how to optionally configure a managed Envoy fleet.
+	Envoy EnvoyConfig `json:"envoy"`
+
+	// Gateway contains parameters for the gateway-api Gateway that Contour
+	// is configured to serve traffic.
+	// +optional
+	Gateway *GatewayConfig `json:"gateway,omitempty"`
+
+	// HTTPProxy defines parameters on HTTPProxy.
+	// +optional
+	HTTPProxy *HTTPProxyConfig `json:"httpproxy,omitempty"`
+
+	// LeaderElection contains leader election parameters.
+	LeaderElection LeaderElectionConfig `json:"leaderElection"`
+
+	// EnableExternalNameService allows processing of ExternalNameServices
+	// Defaults to disabled for security reasons.
+	// +optional
+	// +kubebuilder:default:false
+	EnableExternalNameService bool `json:"enableExternalNameService,omitempty"`
+
+	// RateLimitService optionally holds properties of the Rate Limit Service
+	// to be used for global rate limiting.
+	RateLimitService RateLimitServiceConfig `json:"rateLimitService,omitempty"`
+
+	// Policy specifies default policy applied if not overridden by the user
+	Policy PolicyConfig `json:"policy,omitempty"`
+
+	// Metrics defines the endpoints Envoy use to serve to metrics.
+	// +kubebuilder:default={address: "0.0.0.0", port: 8000}
+	Metrics MetricsConfig `json:"metrics"`
+}
+
+// XDSServerType is the type of xDS server implementation.
+type XDSServerType string
+
+const ContourServerType XDSServerType = "contour"
+const EnvoyServerType XDSServerType = "envoy"
+
+// XDSServerConfig holds the config for the Contour xDS server.
+type XDSServerConfig struct {
+	// Defines the XDSServer to use for `contour serve`.
+	// +kubebuilder:default=contour
+	// +kubebuilder:validation:Enum=contour;envoy
+	Type XDSServerType `json:"type"`
+
+	// Defines the xDS gRPC API address which Contour will serve.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:default="0.0.0.0"
+	Address string `json:"address"`
+
+	// Defines the xDS gRPC API port which Contour will serve.
+	// +kubebuilder:default=8001
+	Port int `json:"port"`
+
+	// Allow serving the xDS gRPC API without TLS.
+	// +kubebuilder:default=false
+	Insecure bool `json:"insecure"`
+
+	// TLS holds TLS file config details.
+	//  +optional
+	TLS *TLS `json:"tls,omitempty"`
+}
+
+// GatewayConfig holds the config for Gateway API controllers.
+type GatewayConfig struct {
+	// ControllerName is used to determine whether Contour should reconcile a
+	// GatewayClass. The string takes the form of "projectcontour.io/<namespace>/contour".
+	// If unset, the gatewayclass controller will not be started.
+	// +kubebuilder:default="projectcontour.io/projectcontour/contour"
+	ControllerName string `json:"controllerName"`
+}
+
+// TLS holds TLS file config details.
+type TLS struct {
+	// CA filename.
+	CAFile string `json:"caFile"`
+
+	// Client certificate filename.
+	CertFile string `json:"certFile"`
+
+	// Client key filename.
+	KeyFile string `json:"keyFile"`
```

Comment:
Should there be defaults here? E.g. https://github.com/projectcontour/contour/blob/main/examples/contour/03-contour.yaml#L43-L45?

Reply:
I don't know how to "turn them off" which is why I didn't provide, meaning, if you want to run `--insecure`, but don't specify the fields, you still get values.

Reply:
I spent a few hours playing with various kubebuilder tags, and confirmed that it works this way.

If you provide defaults inside a struct that has `+optional` and `omitempty`, then those fields will only be used if you specify the struct at all - that is, if it's not empty. But, that does mean that there's no way to default them usefully, using only the kubebuilder defaults, in this `--insecure` case.

So, here, I think we should set the defaults to what Contour *will* default them to, even if it's not used (that is, the values used in the example deployment).

Reply:
Just to clarify @youngnick, even if we set `insecure: true` you'd still get defaults for the keys (ca, cert, key) even though they aren't used?

Reply:
The defaults won't get used unless at least the struct is non-nil. Which would mean that, in YAML form, it's unspecified, which I would guess would be the case in this one. So, if we set defaults *inside* the struct, they won't be used as long as we don't set a default at the outer level.